Add TRD: Tock Code Tiers (#4694)
tock/tock: doc/reference/trd-tiers.md

review as well as user trust in Tock's security. For reviewers and
contributors alike, it can help guide and prepare contributors for the
level of scrutiny their contributions might receive. For example, if a
contribution changes code marked "Critical," it should be clearer that
such changes will require more scrutiny than code marked
"Experimental." Similarly, explicit annotations might deter
contributors from _unnecessarily_ modifying code in higher trust tiers
in contributions that are otherwise unrelated. For users, it can help
inform which subsystems are the most well-scrutinized and tested and
which are unwise to rely on without further auditing.

Explicit annotations in the code itself can also enable tools to
enforce related rules. For example, functions in highly scrutinized
code should not call functions in experimental code.

2 Tiers
===============================

Tock code is grouped into five tiers, with higher numbered tiers denoting more
important code.

| Tier # | Tier         | Description                                                     |
|--------|--------------|-----------------------------------------------------------------|

> | 5      | Verified     | Formally verified code with necessary proof                     |

Reviewer: I don't think "verified" is a reasonable hierarchical "security tier", esp. in comparison with the other tiers on this list.

Formal verification is one particular family of tools we can use to get assurance that our code is doing what it's supposed to, among many other types of tools. Just because something is verified doesn't mean that it's automatically important to the Tock security model. Conversely, just because something is not formally verified does not mean it is not an important component. Some things are just not (yet) tractable to be fully verified. And even then, the particular specification and guarantees matter.

I'd be much more comfortable with removing `Verified` from this list of tiers, and add progressively more encouragement for verification as a tool to gain high assurance to the other, higher tiers.

Author: In a practical sense, if a portion of Tock code is formally verified, even if that Tock code is on the periphery, would a change be expected to also be formally verified? I can imagine that going from formally verified to not formally verified would be a significant change with significant discussion. In that case, annotating that code as such is important.

Author: I do think it is important these convey the difficulty of changing the code.

I have probably been assuming "Verified" really means, in practice, "Critical _and_ Verified". I think a strawman like verifying a obscure temperature sensor capsule illustrates the problem with that assumption, as while that would be verified it doesn't help us much to have that in the top trust tier.

I still think we should aim to have a tier above critical. I'll think more about this.

Code Tiers
========================================

**TRD:** <br/>
**Working Group:** Kernel<br/>
**Type:** Documentary<br/>
**Status:** Draft <br/>
**Author:** Brad Campbell, Amit Levy<br/>
**Draft-Created:** 2025/12/15 <br/>
**Draft-Modified:** 2025/12/15 <br/>
**Draft-Version:** 1 <br/>
**Draft-Discuss:** devel@lists.tockos.org<br/>

Abstract
-------------------------------

As an operating system, Tock has numerous subsytems to support a variety of
hardware platforms, kernel features, and use cases. With a continuously growing
code base, different code modules and files receive a different level of
scrutiny and have different levels of significance for the overall project.
However, this scrutiny is implicit to code authors and reviewers. Code Tiers
addresses this, by adding annotations in the source directly to denote the
importance and scrutiny applied to each code module. This helps clarify to
contributors the expectations for changes to a particular module and signals to
reviewers the level of scrutiny that should be applied to those changes. This

> document is in full compliance with[TRD1][TRD1].

Reviewer 2: ```suggestion
document is in full compliance with [TRD1][TRD1].
```

  tested.

- **Normal**: This is the default tier for all Tock code not otherwise
  classified into another tier.

- **Experimental**: This code is new, a work-in-progress, or otherwise an
  experimental subsystem or module within Tock. It likely will change
  substantially and may be not fully implemented or working. New
  contributions will not receive significant scrutiny to support rapid
  development.

3 Annotation Mechanism
===============================

Code is assigned a tier in Tock using...

3.1 Default Tier
-------------------------------

If there is no annotation present code is, by default, in the Normal
tier.

4 Using the Code Tier Annotations
===============================

Annotating code tiers enables some automated processes to aid Tock development.
This list is not comprehensive, but outlines some anticipated benefits of
annotating code tiers.

1. Detecting code which relies on a lower tier. Code in the highest tiers
   (i.e., critical and verified) may have an implicit assumption that it only
   uses or calls code in the same or higher tier. As an extreme example,
   critical code should not rely on experimental code for normal operation. With

>    annotated code, static analysis can determine if there are any possible

Reviewer 2: I wonder if we can bake this analysis into the Makefile

contribution changes code marked "Critical," it should be clearer that
such changes will require more scrutiny than code marked
"Experimental." Similarly, explicit annotations might deter
contributors from _unnecessarily_ modifying code in higher trust tiers
in contributions that are otherwise unrelated. For users, it can help
inform which subsystems are the most well-scrutinized and tested and
which are unwise to rely on without further auditing.

Explicit annotations in the code itself can also enable tools to
enforce related rules. For example, functions in highly scrutinized
code should not call functions in experimental code.

2 Tiers
===============================

Tock code is tiered based on two dimensions:

1. **Assurance**: How proven, tested, or verified the code is.
2. **Importance**: How critical the code is to the Tock project or its threat
   model.

These dimensions capture the underlying significance of any piece of Tock code,
revealing its relevance to the project and expectations for correctness.

2.1 Assurance
-------------------------------

The assurance tier describes the status of how the code has been shown to be

> correct. This indicates to what extant the code should be considered working,

Reviewer 3: ```suggestion
correct. This indicates to what extent the code should be considered working,
```

and what the expectation is for edits to the code.

| Tier # | Tier                | Description                                        |
|--------|---------------------|----------------------------------------------------|
| 1      | Formally Verified   | Code is checked by formal verification tools       |
| 2      | Extensively Tested  | Code is tested in CI and extensively used          |

> | 3      | Functionally Tested | Code is tested in CI or other automated checks     |

Reviewer 2: Isn't every commit tested in CI before merge, making this the `Normal` tier?

| 4      | Normal              | Code is tested for correctness when it was written |

2.1.1 Assurance Tier Descriptions
-------------------------------

- **Formally Verified**: This code is checked by a static verification tool
  with the necessary annotations and proofs included in the Tock source
  code. This is likely accompanied by extensive test cases, including

>   on-hardware tests. Any changes must include updated checks, such as the
>   necessary annotations and proofs or new tests.

Reviewer 3: ```suggestion
  on-hardware tests. Any changes must include updated checks, such as the
  necessary annotations and proofs or new tests — or they must downgrade
  the assurance tier.
```

To the discussion thread of "a formally verified garbage temperature sensor implementation isn't useful", there should be some kind of escape hatch here.

> - **Extensively Tested**: This code is tested with a combination of Rust unit
>   tests, integration tests, and automatic CI tests. The code is also widely

Reviewer 4: I think this wording is more clear/less ambiguous:

```suggestion
- **Extensively Tested**: This code is tested by an automatic CI system using a combination of Rust unit
  tests, integration tests, and on-hardware tests. The code is also widely
```

Reviewer 4: That said, it's a little hard to see where the lines between "unit test" and "integration test" as well as between "integration test" and "on-hardware" test are.

Does a unit test in a chip crate that tests chip code that depends on kernel code count as an integration test for the kernel code? By some definitions, yes, others no.

Maybe the line here is on-host tests (unit tests and similar), emulated tests (QEMU), and on-hardware tests? IDK how Miri would fit in though, it's generally used to run unit tests but can also emulate target systems.

>   used over a significant time (i.e., multiple years) on multiple hardware

Reviewer 3: I'm wary about including longevity in the assurance metric. When does "new" code become normal; when is normal a "long time"; and, most importantly, what level of change "resets the clock" in a way that would possibly require lowering tier?

I get the point: Code that's been exercised for a long time is more likely correct, but it feels like tracking that would get really fuzzy in practice :/

Reviewer 2: I'm thinking of the release versions. It does take a while for Tock to release new stable versions (2.0 to 2.1 taking a year-ish, and 2.1 to 2.2 taking 3-ish years). 

When a change is made that extensively affects the underlying code, it can be downgraded until the next major release, or an agreed upon sub-version increment depending on how critical the change is.

  platforms by multiple users. This code is highly scrutinized during any
  proposed changes. Extensive testing may be required to validate changes.

> - **Functionally Tested**: This code is tested with some combination of Rust
>   unit tests, integration tests, and automatic CI tests. Changes must pass

Reviewer 4: ```suggestion
- **Functionally Tested**: This code is tested by an automatic CI system using some combination of Rust
  unit tests, integration tests, and on-hardware tests. Changes must pass
```

  or update existing tests, and new tests may be required for any new

>   code.

Reviewer 2: Isn't every commit tested in CI before merge, making this the `Normal` tier?

Reviewer 4: I think the intent of this definition (and the definition of Extensively Tested) above is that "automatic CI tests" refers to on-hardware tests that are automatically run by CI (i.e. Treadmill). If I'm right about that intention, then code that lacks any test cases would not count as Functionally Tested.

But the wording is unclear here, so I'll send some suggestions to update it as well.

- **Normal**: This is the default tier for all Tock code not otherwise
  classified into another tier. The code was tested by running it in an

>   appropriate context and verifying expected behavior.

Reviewer 2: I can see the `Normal` tier being used to draw attention to new pieces of code that were never tested against CI before (as in an open PR), with the tier getting upgraded to functionally tested on merge. 
Unless it means that the developer has to demonstrate that the new code works with some example app?

| Tier # | Tier         | Description                                        |
|--------|--------------|----------------------------------------------------|
| 1      | Critical     | Code is central to Tock and its threat model       |
| 2      | Widely Used  | Code is widely used in most Tock instances         |
| 3      | Normal       | Code is a normal part of the Tock project          |
| 4      | Experimental | Code is new, unfinished, and/or being designed     |

2.2.1 Importance Tier Descriptions
-------------------------------

- **Critical**: This code is directly related to Tock's system-level security
  guarantees. Correctness is critical for Tock to uphold its threat model.
  This code is highly scrutinized during any proposed changes. Changes
  often require extensive discussions and careful audits.

- **Widely Used**: This code is not necessarily related to Tock's security
  model, but is widely used and tested, is necessary for correct operation
  of many platforms, and has been carefully implemented. Changes must be
  extensively tested.

- **Normal**: This is the default tier for all Tock code not otherwise
  classified into another tier.

> - **Experimental**: This code is new, a work-in-progress, or otherwise an
>   experimental subsystem or module within Tock. It likely will change
>   substantially and may be not fully implemented or working. New
>   contributions will not receive significant scrutiny to support rapid
>   development.

Reviewer 3: It feels like too many words here, but I feel like we need to add something somewhere that talks about how code gets out of `Experimental`. Maybe it's as easy as

```suggestion
- **Experimental**: This code is new, a work-in-progress, or otherwise an
  experimental subsystem or module within Tock. It likely will change
  substantially and may be not fully implemented or working. New
  contributions will not receive significant scrutiny to support rapid
  development. When an implementation is complete and churn has settled,
  it should progress to Normal tier. Mainline Tock is happy to carry Experimental
  code for a reasonable, but limited window of time (see https://book.tockos.org/doc/design#merge-aggressively-archive-unabashedly ).
```

Reviewer 2: How do you define complete and churn? 

Maybe we ensure that experimental features don't make their way to stable releases, and define _completeness_ such that upon achieving it, the importance tier can be upgraded in the next release/sub-release based on the importance level of the code (similar to what I said in assurance).

3 Annotation Mechanism
===============================

Code is assigned tiers in Tock using comments on the source code, leveraging
the same infrastructure as doc comments and safety comments.

Any code in Tock that can be commented with a doc comment can be annotated
with tiers.

The code's tier is specified with a top-level header `Code Tier`. Each tier is
specified as a list with the tier name. The format is like this:

```
/// # Code Tier
///
/// - Assurance: <Assurance tier>
/// - Importance: <Importance tier>
///
/// <Optional description or comments>

> ```

Reviewer 4: Does code always have to supply both, or can you supply one and let the other be inherited? I.e.:

```rust
/// # Code Tier
///
/// - Assurance: Normal
/// - Importance: Normal
mod foo {
    // [...]

    // Is omitting Importance here ok?
    /// # Code Tier
    ///
    /// - Assurance: Formally Verified
    #[flux(...)]
    pub fn increment_u32(before: u32) -> u32 {
        before.wrapping_add(1)
    }
}
```

Author: Good question!

Having it fallback to the same thing as though there was no annotation at all seems reasonable. It might also help in the case where it's hard to assign both, and if both are required, then the annotation just gets skipped entirely. However, it might also mean that cases that should have been annotated with both don't get labeled.

Probably ok to allow only one of the annotations.

Reviewer 3: "Inheriting" tiers raises an interesting ergonomics question perhaps. If a new method were added to `mod foo` here, without any annotations, then it would implicitly inherit the 'parent' tiers, which feels hazardous — I suppose the tooling could simply enforce that "if an outer block has an explicit tier, then any new inner elements must have explicit tiers"?

Or really I guess my question is: should we have an module-level tier at all? Or is `mod foo`'s tier literally the sum of its parts? What does it mean to have e.g. an `Extensively Tested` module with a new(?) `Normal` method added?

Reviewer 4: I think we have to decide between:

1. Using inheritance, and having that sometimes be inaccurate (the drawback Pat mentioned).
1. Using a lot of copy + paste, and having those annotations bitrot.

In the long run, I think they'll end up having roughly the same level of accuracy, but inheriting tiers seems like less work. I certainly don't want to see us add a 4-line tier comment to every function in the codebase.

My guess is that we'll end up annotating core crates (kernel and arch crates) at roughly the module level, and board/chip crates at roughly the crate level. We can get more detailed over time (and perhaps push in that direction by tightening the restrictions on what-code-can-call-what), but that seems like the practical starting point.

For example:

```rust
/// Create a new Tock process.
///
/// # Code Tier
///
/// - Assurance: Critical
/// - Importance: Extensively Tested

> pub fn process_standard_create() -> ProcessStandard;

Reviewer 4: As an alternative to using doc comments, we could create attribute macros to specify the code's tier, something like:

```rust
/// Create a new Tock process.
#[tock_assurance(Critical)]
#[tock_importance(ExtensivelyTested)]
pub fn process_standard_create() -> ProcessStandard;
```

These proc macros could validate that the assurance/importance levels passed to them are valid, but otherwise are no-ops (they just pass their inputs through to their output). I believe they could easily be written without `syn`/`quote`/`proc_macro2`. We'd still use an external tool to run whatever analyses we want to run (e.g. tagging PRs based on the code tier they touch or erroring if e.g. critical code calls experimental code).

There are a couple advantages to using attribute macros over doc comments:

1. Load-bearing comments is weird and unintuitive for new contributors.
2. Moving these into a macro removes the "looks like prose" and "is valid Markdown" contraints, letting us pick a simpler syntax that is likely easier to write.
3. We can choose to validate the annotations during a normal `cargo check`, rather than waiting until you run the check-tock-tier-annotations (name to be bikeshed) tool to emit those errors.

There are some drawbacks to the attribute macros:

1. They can't be used everywhere that doc comments can. A full list of allowed positions is in [the reference](https://doc.rust-lang.org/reference/procedural-macros.html#r-macro.proc.attribute.use-positions).
2. Writing our validation tool could be awkward with `use` statements. I hit this problem with the expand_macros tool in tock-registers. It does not understand `use` statements, and is instead hardcoded to search for `register_map`, `tock_registers::register_map`, or `::tock_registers::register_map`. If someone does `use assurance = tock_assurance;` then the tool might be confused (depends on what tool we use to parse/understand the Rust code).

Also, a key tradeoff is whether we want these to show up in our rustdoc, and I'm not sure which is better in that respect.

Just tossing this out in case using an attribute macro hasn't been considered before.

Author: We did consider this, and we certainly haven't ruled it out. Extracting comments just seemed easier/more straightforward to get started with.

```

3.1 Default Tier
-------------------------------

If there is no annotation present code is, by default, in the Normal assurance

> tier and the Normal importance tier.

Reviewer 3: ```suggestion
tier and the Normal importance tier.

Newly authored code is not always `Experimental`. Often it may be appropriate,
especially for smaller changesets or simple fixes from trusted contributors, for
code to enter the tree at `Normal` tier, or possibly even higher tiers if it enters
with extensive testing or verification apparatus.
```

4 Using the Code Tier Annotations
===============================

Annotating code tiers enables some automated processes to aid Tock development.
This list is not comprehensive, but outlines some anticipated benefits of
annotating code tiers.

1. Detecting code which relies on a lower tier. Code in the highest assurance
   tiers(i.e., Formally Verified and Extensively Tested) may have an implicit
   assumption that it only uses or calls code in the same or higher tier.

>    Similarly, code in a high importance tier may not expect to depend on code in

Reviewer 2: I am unsure if this may not is talking about "is not allowed"; or it not expected tohappen, but may happen.

   a low importance tier. As an extreme example, code in the Critical importance

>    tier should not rely on code in the Experimental importance tier. With

Reviewer 2: And as non-extreme examples, would we be open to adjacent tiers being allowed? For example, Widely-Used may depend on Normal, given that this distinction seems to be a function of time/adoption rather than the importance to the threat model in general (if not, that is unclear to me).

If so, would we word this to say something along the lines of "Higher tier code _may_ depend on code that is one tier lower except in the case of Critical importance codes, which requires there be no dependence on lower tier importance code. Lower tier code, meanwhile can depend on higher tier code."? 

I feel like this is a _somewhat_ easier delineation for Assurance tiers, but less so for importance tiers.

   annotated code, static analysis can determine if there are any possible

>    execution paths where any code relies on lower-tier code.

Reviewer 4: No need to address this or change anything, just pointing out something we'll have to think about at some point.

Now that Assurance and Importance are separate, it is no longer obvious when one bit of code is higher-tier than another bit of code. Like, code can be (Formally Verified, Experimental), is that higher or lower tier than (normal, normal) code?

We'll probably want to annotate some code and then work out what sort of dependencies we're okay with.

Reviewer 2: Intuitively, Importance comes before Assurance. Formally verified is what we strive for, however, there could be something about a program that makes it important despite not  being verified.
So despite code being formally verified, if it is incomplete for any reason, making it experimental, I would still rank it below code that is of widely-used importance level, but is not formally verified. 

That said, I am not entirely sure this assumption will always hold true, so I agree with checking with a set of samples.

>    With Tock's modularity, it is possible that certain kernel configurations

Reviewer 2: I feel like this sentence would go better after the sentence on line 205.
```
Similarly, code in a high importance tier may not expect to depend on code in a low importance tier.
``` 
For a while I wasn't sure if the _may_ in the above sentence meant "should not be allowed", or if it meant "ideally should not happen, but may happen". 

However, reading this sentence clarified that it was the latter.

   would unexpectedly cause high tier code to rely on low tier code. This
   analysis would help Tock developers and users detect these scenarios.

2. Code review assistance. Reviewing Tock PRs does not require the same scrutiny
   across all code. However, the required scrutiny is typically determined
   based on experience and intuition. With labeled code tiers, the expectation

>    for code review is made explicit.

Reviewer 2: I think what is not clear to me is how a developer adding/modifying code trusted to use the correct tags. If we are going to base the review scope based on the tags, but the tags are lying, how do we resolve that? Is the CI pipeline somehow supposed to verify the tag before a pull-request passes the checks? I think this is more relevant for the `Importance` tag.

For example, what is stopping a user from adding a new critical change and marking it as `Normal`? Do all new changes start off as experimental in terms of importance and only get upgraded with adoption scale over time? The way I read it, the `Experimental` tier suggests any new code gets tagged with the `Experimental` tag.

3. Avoiding undetected changes to trusted code. A contribution may change code
   that is in the normal tier, but that code is used by code in a higher tier.
   Control flow analysis from code in higher tiers would reveal the potential
   impact of this change and mark the contribution for increased scrutiny.

5 Open Questions
===============================

1. Is higher tier code relying on lower tier code an error or a warning? If a
   warning, how do we specify the cases where we want it to be an error? For
   example, we may mark a function as a high tier, and then examine all code it
   depends on to ensure it is at the same tier or higher. We then want to be
   able to record that, so any future change that makes that no longer hold is
   flagged. If an error, how do we specify the cases that we don't want to cause

>    CI to fail?

Reviewer 4: IMO, if CI ignores warnings then there's not much point to having the warning. If warnings are treated as errors, there isn't a distinction.

I do think we should be able to compile code that breaks the tier policy for development reasons (i.e. you are still reworking things), but if we have any automated enforcement then `make prepush` should fail if tiers are used incorrectly.

Reviewer 3: I think mechanically emitting a warning (and having CI deny warnings) is the right solution here — but we will need an escape hatch that allows (silences) the warning for cases where it's appropriate.

I imagine we'd annotate the call site, looking something like this:

```rust
#[tock_assurance(Critical)]
#[tock_importance(ExtensivelyTested)]
pub fn process_standard_create() -> ProcessStandard {
  ...
  // ~~Safety~~ Importance: We downgrade here because...
  #[tock_downgrade_importance(Experimental)]
  arch_helper_we_need_but_dont_like_or_whatever();

  ...
  // Assurance: We downgrade here because...
  #[tock_downgrade_assurance(Normal)]
  pretty_printer_for_debug_that_is_useful_but_modifies_no_state();
```

Where _that function call_ (and its transients) allow lower tiers
